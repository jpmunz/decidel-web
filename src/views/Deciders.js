import React from "react";
import { connect } from "react-redux";
import { getCurrentUserName, getDeciders } from "src/state";
import { Button, InputGroup, FormControl } from "react-bootstrap";
import autoBind from "react-autobind";
import "./Deciders.scss";
import arrow from "src/assets/arrow.svg";

export class Deciders extends React.Component {
  constructor(props) {
    super(props);
    autoBind(this);
  }

  handleURLCopy() {
    navigator.clipboard.writeText(window.location.href);
  }

  renderDecider(decider, i) {
    let turnIndicator;
    let classNames = [];

    if (decider.isNext) {
      turnIndicator = (
        <img className="turn-indicator" src={arrow} alt="current turn"></img>
      );
      classNames.push("current-turn");
    }

    let name = decider.name;
    if (decider.name === this.props.currentUserName) {
      classNames.push("current-user");
      name = "You";
    } else {
      classNames.push("other-user");
    }

    return (
      <li key={i}>
        {turnIndicator}
        <span className={classNames.join(" ")} title={decider.name}>
          {name}
        </span>
      </li>
    );
  }

  render() {
    return (
      <div className="Deciders">
        <div className="deciders-list-container">
          <ul>{this.props.deciders.map(this.renderDecider)}</ul>
        </div>

        <div className="deciders-share">
          <label htmlFor="copy-decidel-url">
            Share this link with your fellow deciders
          </label>
          <InputGroup className="mb-3">
            <FormControl
              aria-label="Decidel URL"
              readOnly={true}
              value={window.location.href}
            />
            <InputGroup.Append>
              <Button
                id="copy-decidel-url"
                variant="outline-primary"
                onClick={this.handleURLCopy}
              >
                Copy
              </Button>
            </InputGroup.Append>
          </InputGroup>
        </div>
      </div>
    );
  }
}

export default connect((state, props) => {
  return {
    deciders: getDeciders(state, props.id),
    currentUserName: getCurrentUserName(state)
  };
})(Deciders);
