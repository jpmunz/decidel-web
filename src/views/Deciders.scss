@import "mixins";

.Deciders {
  .deciders-list-container {
    width: 200px;
    @include rounded();
    @include box-shadow(10px, 10px, 20px, rgba(0, 0, 0, 0.2));
  }

  ul {
    width: 60%;
    margin: 0 auto;

    span {
      cursor: default;
      @include ellipsis();
    }
  }

  .turn-indicator {
    width: 20px;
    vertical-align: top;
  }

  .current-user {
    font-weight: bold;
  }

  .other-user,
  .current-user {
    &:not(.current-turn) {
      margin-left: 25px;
    }
  }

  .current-turn {
    margin-left: 5px;
  }

  .deciders-share {
    margin-top: 30px;
  }
}
