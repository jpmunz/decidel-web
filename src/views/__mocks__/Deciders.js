import React from "react";

export const MockDeciders = jest.fn();

MockDeciders.mockReturnValue(<div>Mocked Deciders</div>);

const mock = jest.fn().mockImplementation(MockDeciders);

export default mock;
