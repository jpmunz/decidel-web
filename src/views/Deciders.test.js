import React from "react";
import { Deciders } from "./Deciders";
import { render } from "@testing-library/react";

const renderComponent = function({
  id = "22",
  deciders = [],
  currentUserName = "Bob"
}) {
  return render(
    <Deciders id={id} currentUserName={currentUserName} deciders={deciders} />
  );
};

const renderTest = options =>
  expect(renderComponent(options).asFragment()).toMatchSnapshot();

test("Current user is next", () =>
  renderTest({
    currentUserName: "Jon",
    deciders: [
      { name: "Bob", isNext: false },
      { name: "Jon", isNext: true }
    ]
  }));
test("Another user is next", () =>
  renderTest({
    currentUserName: "Bob",
    deciders: [
      { name: "Bob", isNext: false },
      { name: "Jon", isNext: true }
    ]
  }));
