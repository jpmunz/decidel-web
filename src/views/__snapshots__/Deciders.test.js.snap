// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Another user is next 1`] = `
<DocumentFragment>
  <div
    class="Deciders"
  >
    <div
      class="deciders-list-container"
    >
      <ul>
        <li>
          <span
            class="current-user"
            title="Bob"
          >
            You
          </span>
        </li>
        <li>
          <img
            alt="current turn"
            class="turn-indicator"
            src="arrow.svg"
          />
          <span
            class="current-turn other-user"
            title="Jon"
          >
            Jon
          </span>
        </li>
      </ul>
    </div>
    <div
      class="deciders-share"
    >
      <label
        for="copy-decidel-url"
      >
        Share this link with your fellow deciders
      </label>
      <div
        class="mb-3 input-group"
      >
        <input
          aria-label="Decidel URL"
          class="form-control"
          readonly=""
          value="http://localhost/"
        />
        <div
          class="input-group-append"
        >
          <button
            class="btn btn-outline-primary"
            id="copy-decidel-url"
            type="button"
          >
            Copy
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Current user is next 1`] = `
<DocumentFragment>
  <div
    class="Deciders"
  >
    <div
      class="deciders-list-container"
    >
      <ul>
        <li>
          <span
            class="other-user"
            title="Bob"
          >
            Bob
          </span>
        </li>
        <li>
          <img
            alt="current turn"
            class="turn-indicator"
            src="arrow.svg"
          />
          <span
            class="current-turn current-user"
            title="Jon"
          >
            You
          </span>
        </li>
      </ul>
    </div>
    <div
      class="deciders-share"
    >
      <label
        for="copy-decidel-url"
      >
        Share this link with your fellow deciders
      </label>
      <div
        class="mb-3 input-group"
      >
        <input
          aria-label="Decidel URL"
          class="form-control"
          readonly=""
          value="http://localhost/"
        />
        <div
          class="input-group-append"
        >
          <button
            class="btn btn-outline-primary"
            id="copy-decidel-url"
            type="button"
          >
            Copy
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
